### PR TITLE
[C++20] [Modules] Introduce reduced BMI

### LINKED ISSUE #71034: [C++20] [Modules] Try to introduce thin BMI to exclude the things not necessary in an interface
Let's call the current BMI for named modules as fat BMI. Since it contains all the information needed to compile the BMI itself to an object file. And this is the foundation of 2 phase compilations.

However, it is a waste of time and space if we're using one phase compilations. 

Then let's call the BMI which doesn't contain all the information to produce an object file, but is able to satisfy the requirement of consumers of the modules as thin BMI. Then it is clear that we should use thin BMI then fat BMI in one phase compilation.

Also in 2 phase compilations, the thin BMI may be helpful too if the build systems want to use a compare-and-swap strategy to generate the BMI based on the BMIDeclsHash (See https://github.com/llvm/llvm-project/issues/71618).

### PR DESCRIPTION
**Pull request by ChuanqiXu9**
Close https://github.com/llvm/llvm-project/issues/71034

See
https://discourse.llvm.org/t/rfc-c-20-modules-introduce-thin-bmi-and-decls-hash/74755

This patch introduces reduced BMI, which doesn't contain the definitions of functions and variables if its definitions won't contribute to the ABI.

Testing is a big part of the patch. We want to make sure the reduced BMI contains the same behavior with the existing and relatively stable fatBMI. This is pretty helpful for further reduction.

The user interfaces part it left to following patches to ease the reviewing.

### CODE REVIEW

**Comment by ChuanqiXu9** [comment]
My impression to the feedbacks is that every one of us loves the direction, while we may need more agreement on the user interfaces.

To make it easier to review, I split all the user interfaces related part to following patches. So that the current patch won't affect users.

This patch introduced an CC1 option `-emit-thin-module-interface` and tests the behavior for almost every places we can to make sure it is stable. While the diff is big, we can find most parts of it are boiler plate, so don't be panic.

I feel the patch good and hope we can land this quickly so that we can start the work to help users actually.

**Comment by ChuanqiXu9** [comment]
@iains @dwblaikie ping~

**Comment by iains** [comment]
@ChuanqiXu9 very sorry for the slow review.  It would help me if the design was described in the commit message instead of trying to deduce it from the patch (maybe it's in a thread somewhere - so a cross-reference would help).

two immediate questions and one observation:

- I see you are using a multiplex consumer (actually, for some reason, I thought you were objecting to that part of my design); does this mean that your proposed solution can emit both the object and the reduced BMI from a single compilation job?

- I was concerned from earlier conversations that this design might require a codegen back end to be instantiated to allow the reduced BMI (which would be bad for --precompile/-fmodule-only type jobs). Any comments?

- It would be better to avoid introducing more layering violations but, as we discussed in face-to-face meetings, I have less concern on the output side.  It still seems to me that the best model is one where we have AST transforms (that very likely need Sema to be correct) and then the serializer is a simple as possible.

so  something like

raw AST  ======== > codegen (when required)
           +
           + =====> AST transforms ====> BMI output.

As I understand the patch you are combining the transform with the output?

**Comment by ChuanqiXu9** [comment]
> @ChuanqiXu9 very sorry for the slow review. It would help me if the design was described in the commit message instead of trying to deduce it from the patch (maybe it's in a thread somewhere - so a cross-reference would help).

hi @iains , sorry for the confusion. It may be hard to balance how detail the message it is. So it is no problem at all to ask questions as long as it need. This should be the point of the review process too.

For this patch itself, the design is to introduce a  CC1 command line `-emit-reduced-module-interface` and the corrresponding action `GenerateReducedModuleInterfaceAction`. Both of them are aimed to help testing reduced BMI instead of being used by end users actually. Then it might be straightforward to get it by seeing the implementation of `GenerateReducedModuleInterfaceAction`.

Like I said in the commit message, this patch itself doesn't involve anything relevant to user interfaces. I left it to the latter patches.

> 
> two immediate questions and one observation:
> 
> * I see you are using a multiplex consumer (actually, for some reason, I thought you were objecting to that part of my design); 

In fact, I am not objecting your design due to multiplex consumer. I am objecting your design in implementing it in `FrontendAction::CreateWrappedASTConsumer()`. I think this is not the correct place to introduce language or module specific things.

> does this mean that your proposed solution can emit both the object and the reduced BMI from a single compilation job?

Yes, this is the goal. I've already had a patch in the downstream to do that. But I am still wondering if I can implement it better.

> * I was concerned from earlier conversations that this design might require a codegen back end to be instantiated to allow the reduced BMI (which would be bad for --precompile/-fmodule-only type jobs). Any comments?

I am not sure if I understand this. What does it mean "require a codegen back end to be instantiated to allow the reduced BMI "? Do you mean to not touch CodeGen part or to not touch the CodeGen action? My local patch touched the code gen action without touching any real CodeGen related things. 

>  the reduced BMI (which would be bad for --precompile/-fmodule-only type jobs)

For `--precompile/-fmodule-only` type jobs, I'll create another action to make it (Similar to existing `GenerateModuleInterfaceAction`). Then both of the actions will try to reuse the same consumer `ReducedBMIGenerator` to avoid repeated works.

> * It would be better to avoid introducing more layering violations but, as we discussed in face-to-face meetings, I have less concern on the output side.  It still seems to me that the best model is one where we have AST transforms (that very likely need Sema to be correct) and then the serializer is a simple as possible.

Yeah, it should be less concerned. BTW, currently the simpler serializer/deserializer should be ASTRecordWriter/ASTRecordReader. And the current ASTReader/ASTWriter takes some semantical job.

> 
> so something like
> 
> ```
> raw AST +======== > codegen (when required)
>         |
>         + =====> AST transforms ====> BMI output.
> ```
> 
> As I understand the patch you are combining the transform with the output?

On the one hand, the **current** patch doesn't do that. The current patch is almost a NFC patch. It belongs to the following patch. On the other hand, the answer may be yes. Probably I did the `AST transforms` you said in the AST Writer. I don't feel it is so awful.

---

Given all of us loves reduced BMI, I suggest we can focus on current patch then discuss user interfaces related things in the next patch after this got landed.

**Comment by iains** [comment]
> Like I said in the commit message, this patch itself doesn't involve anything relevant to user interfaces. I left it to the latter patches.

Are you in a position to post the next patch (at least as a draft)?  That would help me see the direction.

> > * I was concerned from earlier conversations that this design might require a codegen back end to be instantiated to allow the reduced BMI (which would be bad for --precompile/-fmodule-only type jobs). Any comments?
> 
> I am not sure if I understand this. What does it mean "require a codegen back end to be instantiated to allow the reduced BMI "? Do you mean to not touch CodeGen part or to not touch the CodeGen action? My local patch touched the code gen action without touching any real CodeGen related things.

> > the reduced BMI (which would be bad for --precompile/-fmodule-only type jobs)
> 
> For `--precompile/-fmodule-only` type jobs, I'll create another action to make it (Similar to existing `GenerateModuleInterfaceAction`). Then both of the actions will try to reuse the same consumer `ReducedBMIGenerator` to avoid repeated works.

OK, that answers my concern (which was that we might have to add the code-gen backend to a --precompile if that was the mechanism used to do the BMI reduction).

> > * It would be better to avoid introducing more layering violations but, as we discussed in face-to-face meetings, I have less concern on the output side.  It still seems to me that the best model is one where we have AST transforms (that very likely need Sema to be correct) and then the serializer is a simple as possible.
> 
> Yeah, it should be less concerned. BTW, currently the simpler serializer/deserializer should be ASTRecordWriter/ASTRecordReader. And the current ASTReader/ASTWriter takes some semantical job.

... and, on the reader side, that already gives us some big problems (as I say, I am less concerned on the writer side, but who can see the whole future?).

> > so something like
> > ```
> > raw AST +======== > codegen (when required)
> >         |
> >         + =====> AST transforms ====> BMI output.
> > ```
> > As I understand the patch you are combining the transform with the output?
> 
> On the one hand, the **current** patch doesn't do that. The current patch is almost a NFC patch. It belongs to the following patch. On the other hand, the answer may be yes. Probably I did the `AST transforms` you said in the AST Writer. I don't feel it is so awful.

Maybe not for the short-term relatively simple tasks - but we should also take a view on the medium and longer term (for example, GMF decl elision is likely to be helpful to users in reducing both the size of the BMI and the number of decls that need merging on input).

We need the AST in this path to be mutable (including removal of decls); that way each transform can be maintained as a separate entity - I think that if we end up doing "many transforms" as part of the output, it will become very confusing.

_(although, to be clear, i**n the short-term we might agree to do the work in the output** - I really do think it would be bad to make that the long term mechanism)._

> Given all of us loves reduced BMI, I suggest we can fosus on current patch then discuss user interfaces related things in the next patch after this got landed.

We do all want to produce the reduced BMI, I agree; but we also always have limited resources to do the work, so that it would be good to try and pick an implementation that will be smooth for the future work too.

I understand the purpose of the current patch better now - and will try to take a more detailed look over the next few days - as noted above, it would help very much to have a preview of the next patch in the series.

**Comment by ChuanqiXu9** [comment]
> > Like I said in the commit message, this patch itself doesn't involve anything relevant to user interfaces. I left it to the latter patches.
> 
> Are you in a position to post the next patch (at least as a draft)? That would help me see the direction.

I post it here https://github.com/ChuanqiXu9/llvm-project/commit/efcc7e888a96e9935a354759de320dea55e93105 since I didn't get how to send stacked pr in github yet. But I guess it might be sufficient since it is really in an early phase.

> 
> > > * I was concerned from earlier conversations that this design might require a codegen back end to be instantiated to allow the reduced BMI (which would be bad for --precompile/-fmodule-only type jobs). Any comments?
> > 
> > 
> > I am not sure if I understand this. What does it mean "require a codegen back end to be instantiated to allow the reduced BMI "? Do you mean to not touch CodeGen part or to not touch the CodeGen action? My local patch touched the code gen action without touching any real CodeGen related things.
> 
> > > the reduced BMI (which would be bad for --precompile/-fmodule-only type jobs)
> > 
> > 
> > For `--precompile/-fmodule-only` type jobs, I'll create another action to make it (Similar to existing `GenerateModuleInterfaceAction`). Then both of the actions will try to reuse the same consumer `ReducedBMIGenerator` to avoid repeated works.
> 
> OK, that answers my concern (which was that we might have to add the code-gen backend to a --precompile if that was the mechanism used to do the BMI reduction).
> 
> > > * It would be better to avoid introducing more layering violations but, as we discussed in face-to-face meetings, I have less concern on the output side.  It still seems to me that the best model is one where we have AST transforms (that very likely need Sema to be correct) and then the serializer is a simple as possible.
> > 
> > 
> > Yeah, it should be less concerned. BTW, currently the simpler serializer/deserializer should be ASTRecordWriter/ASTRecordReader. And the current ASTReader/ASTWriter takes some semantical job.
> 
> ... and, on the reader side, that already gives us some big problems (as I say, I am less concerned on the writer side, but who can see the whole future?).

I guess you're referring the process how we decide a declaration is visible?

> 
> > > so something like
> > > ```
> > > raw AST +======== > codegen (when required)
> > >         |
> > >         + =====> AST transforms ====> BMI output.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > As I understand the patch you are combining the transform with the output?
> > 
> > 
> > On the one hand, the **current** patch doesn't do that. The current patch is almost a NFC patch. It belongs to the following patch. On the other hand, the answer may be yes. Probably I did the `AST transforms` you said in the AST Writer. I don't feel it is so awful.
> 
> Maybe not for the short-term relatively simple tasks - but we should also take a view on the medium and longer term (for example, GMF decl elision is likely to be helpful to users in reducing both the size of the BMI and the number of decls that need merging on input).

For GMF decl elision, I posted a patch to implement it in ASTWriter and I reposted it in https://github.com/llvm/llvm-project/pull/76930. The big problem is that this is not formal. (Just for sharing, I am not proposing this now)

> 
> We need the AST in this path to be mutable (including removal of decls); that way each transform can be maintained as a separate entity - I think that if we end up doing "many transforms" as part of the output, it will become very confusing.

While the model sounds good, I am pessimistic for making it correctly, completely, and efficiently.

> 
> _(although, to be clear, i**n the short-term we might agree to do the work in the output** - I really do think it would be bad to make that the long term mechanism)._

Not against, I just think it is not so bad. There already many optimizations in the current serializations.

> 
> > Given all of us loves reduced BMI, I suggest we can fosus on current patch then discuss user interfaces related things in the next patch after this got landed.
> 
> We do all want to produce the reduced BMI, I agree; but we also always have limited resources to do the work, so that it would be good to try and pick an implementation that will be smooth for the future work too.

Understood. I just think it won't be too bad. Or it is not easy for us to get a much better solution in the resources we have. I prefer the style that don't make perfect the enemy of better.

> 
> I understand the purpose of the current patch better now - and will try to take a more detailed look over the next few days

I don't intend to land this in clang18. So we don't need to be hurry.

> - as noted above, it would help very much to have a preview of the next patch in the series.

Sent. But I just feel it is not so helpful for reviewing this patch : )

**Comment by ChuanqiXu9** [comment]
@mizvekov would you like to take a look at this? This is related (in some level) to what you say https://github.com/llvm/llvm-project/pull/79959. In short, in the direction, in the one-phase compilation, the pcm won't be compiled into object files. (But it still will in the two-phase compilation).

**Comment by ChuanqiXu9** [comment]
@iains @mizvekov ping~

**Comment by iains** [comment]
* I do not want to block progress, so let's move forward with this patch for now.

* It seems to me (as we found with GMF decl elision) that the process is quite a bit more complex than simply omitting a decl.  We need to elide other decls that are then unused (e.g. decls local to an elided function body) and also avoid emitting types that are now no longer existent.

The process seems to me to be either one of:
 - rewriting the AST (which is why my patch set picked the use of the plugin API since that is the purpose there).
 - walking the AST and marking entities as used / not used / elided.
 
 It still feels to me to be better to have clear separation of this work from the work of streaming - but if we can make clear layers within the streaming, then maybe the maintenance will not be too hard.

**Comment by ChuanqiXu9** [comment]
> * I do not want to block progress, so let's move forward with this patch for now.

Yeah. Great to see we have some progress finally. I think this is really important since I see more and more peope complaninig the performance for modules. I feel this series patch is key to to improve the user's impression that named modules are just purely wrappers for PCH.

> * It seems to me (as we found with GMF decl elision) that the process is quite a bit more complex than simply omitting a decl.  We need to elide other decls that are then unused (e.g. decls local to an elided function body) and also avoid emitting types that are now no longer existent.

After looking into the codes more, I changed my mind for this. I feel it is the most efficient and the most natural way to omit declarations in ASTWriter.

The idea is, previously, we'll start to write declarations in the current TU from the first declarations. And in the C++20 named modules world, we would start to write declarations from the first declarations in the global module fragment.  And we can implement GMF decl elision naturally by writing declarations from the first declaration in the module purview and only write the referenced decl from the GMF during the writing process. This is super natural and efficient.

> 
> The process seems to me to be either one of:
> 
> * rewriting the AST (which is why my patch set picked the use of the plugin API since that is the purpose there).
> * walking the AST and marking entities as used / not used / elided.

Now I doubt both of the method to be not efficiency and it adds additional burdens to the developers and the users.

> 
> It still feels to me to be better to have clear separation of this work from the work of streaming - but if we can make clear layers within the streaming, then maybe the maintenance will not be too hard.

I think the maintainance may not be too hard in that way. ASTWriter is not such a devil :) Probably we need some refactoration in the serialization to make codes more clear. But the point is that we must get the ASTWriter/ASTReader involved. It may not be a good idea to leave the ASTWriter/ASTReader as is and add another layer... 

For example, it should be better to not deserialize the declaration from the very beginning instead of deserializing the declaration and judge that it is not wanted/visible. And to achieve this, we must  touch the serialization layer deeply.

File: clang/include/clang/Serialization/ASTWriter.h [hunk 1]
```diff
@@ -830,6 +843,19 @@ class PCHGenerator : public SemaConsumer {
...
+  ReducedBMIGenerator(const Preprocessor &PP, InMemoryModuleCache &ModuleCache,
+                      StringRef OutputFile, std::shared_ptr<PCHBuffer> Buffer,
+                      bool IncludeTimestamps);
+
+  void HandleTranslationUnit(ASTContext &Ctx) override;
+};
+
+/// If the definition may impact the ABI. If yes, we're allowed to eliminate
+/// the definition of D in reduced BMI.
+bool MayDefAffectABI(const Decl *D);
```
[the remark this answers is not in the document]
Comment (ChuanqiXu9) [reply clang/include/clang/Serialization/ASTWriter.h]: Nice catch up.

I changed the name to `CanElideDeclDef` and the comments to:

```
/// If we can elide the definition of \param D in reduced BMI.
///
/// Generally, we can elide the definition of a declaration if it won't affect the
/// ABI. e.g., the non-inline function bodies.
```

Hope this to be clear.

**Comment by iains** [comment]
Do you expect to make any changes to type streaming?

**Comment by ChuanqiXu9** [comment]
> Do you expect to make any changes to type streaming?

I don't expect to do that explicitly. The number of types deserialized can be decreased naturally after we avoid emitting declarations during the writing.

**Comment by ChuanqiXu9** [comment]
I am going to land this in the week later if no objections come in. I think it is necessary to land the series of patches (to reduce the contents of BMI) for clang19. And of course, the functionality will be opt in for one~two releases for experimental.

**Comment by iains** [comment]
I have no further comments so LGTM if there are no objections from the other reviewers this week.

**Comment by cor3ntin** [comment]
Can we add a release note and documentation for this?
Thanks!

**Comment by ChuanqiXu9** [comment]
> Can we add a release note and documentation for this? Thanks!

The current patch is transparent to users and it is only part of the series patches. I'd like to document that after I made the series of patches.

**Pull request merged**